guidance/models/_azure_guidance.py

```diff
@@ -0,0 +1,167 @@
+import requests
+import os
+import base64
+import json
+import urllib.parse
+from ._model import Engine, Model, EngineCallResponse
+from .._chat import Phi3ChatTemplate
+from ._byte_tokenizer import ByteTokenizer
+
+
+class AzureGuidanceEngine(Engine):
+    """This connects to a remote guidance server on Azure and runs all computation using the remote engine."""
+
+    def __init__(self, server_url, max_streaming_tokens=1000, chat_template=None):
+        if (
+            server_url is None
+            or isinstance(server_url, str)
+            and len(server_url.strip()) == 0
+        ):
+            server_url = os.getenv("AZURE_GUIDANCE_URL", "")
+        elif not isinstance(server_url, str):
+            raise ValueError("server_url must contain a URL string.")
+
+        if not server_url.startswith("http"):
+            raise ValueError(
+                "AzureGuidance requires a remote model URL that starts with http"
+            )
+        self.server_url = server_url
+        self.max_streaming_tokens = max_streaming_tokens
+
+        if chat_template is None:
+            # TODO [PK]: obtain this from the server
+            chat_template=Phi3ChatTemplate
+
+        tokenizer = ByteTokenizer(chat_template)
+
+        # build the Engine
+        super().__init__(tokenizer=tokenizer, compute_log_probs=False)
+
+    def __call__(self, parser, grammar, ensure_bos_token=True):
+        b64 = base64.b64encode(grammar.serialize()).decode("utf-8")
```

Comment:
Side note (and definitely not for now): I wish we used JSON, not protobuf. I doubt the performance difference is meaningful compared to an LLM forward pass.

Reply:
I mostly agree and would prefer JSON if possible.  There are some unresolved questions though about whether the grammars can get unwieldy in JSON.  In theory the grammars could get much larger than the prompt text.